### PR TITLE
feat: implemented custom errors

### LINKED ISSUE #45: Implement proper python exceptions for our different error cases
For now, we only have `RuntimeError`, which is clearly sub-optimal

### PR DESCRIPTION
**Pull request by lukapeschke**
This adds custom exception types, which should make error management by users easier

closes #45

### CODE REVIEW

File: python/fastexcel/__init__.py [hunk 1]
```diff
@@ -12,7 +12,18 @@
 
 import pyarrow as pa
 
-from ._fastexcel import __version__, _ExcelReader, _ExcelSheet
+from ._fastexcel import (
+    ArrowError,
+    CalamineCellError,
+    CalamineError,
```
Comment (PrettyWood) [inline python/fastexcel/__init__.py:18]: Could we add docstrings? The doc could hence explain in which case those can happen

File: src/types/excelreader.rs [hunk 2]
```diff
@@ -20,9 +23,12 @@ pub(crate) struct ExcelReader {
 impl ExcelReader {
     // NOTE: Not implementing TryFrom here, because we're aren't building the file from the passed
     // string, but rather from the file pointed by it. Semantically, try_from_path is clearer
-    pub(crate) fn try_from_path(path: &str) -> Result<Self> {
-        let sheets = open_workbook_auto(path)
-            .with_context(|| format!("Could not open workbook at {path}"))?;
+    pub(crate) fn try_from_path(path: &str) -> FastExcelResult<Self> {
+        let sheets = match open_workbook_auto(path) {
+            Ok(ok) => Ok(ok),
+            Err(err) => Err(FastExcelErrorKind::CalamineError(err).into()),
```
Comment (PrettyWood) [inline src/types/excelreader.rs:29]: map_err ?

File: src/types/excelreader.rs [hunk 3]
```diff
@@ -55,14 +61,16 @@ impl ExcelReader {
         skip_rows: usize,
         n_rows: Option<usize>,
         schema_sample_rows: Option<usize>,
-    ) -> Result<ExcelSheet> {
+    ) -> PyResult<ExcelSheet> {
         let range = self
             .sheets
             .worksheet_range(&name)
-            .with_context(|| format!("Error while loading sheet {name}"))?;
+            .map_err(|err| FastExcelErrorKind::CalamineError(err).into())
```
Comment (PrettyWood) [inline src/types/excelreader.rs:68]: What I meant with my previous comment.

File: src/error.rs [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+    // Calamine error
+    create_exception!(_fastexcel, CalamineError, FastExcelError);
+    // Sheet not found
+    create_exception!(_fastexcel, SheetNotFoundError, FastExcelError);
+    // Arrow error
+    create_exception!(_fastexcel, ArrowError, FastExcelError);
+    // Invalid parameters
+    create_exception!(_fastexcel, InvalidParametersError, FastExcelError);
+
+    pub(crate) trait IntoPyResult {
```
Comment (PrettyWood) [inline src/error.rs:148]: TryInto since it returns a result?

**Review by PrettyWood — COMMENTED** [review]
Overall looks great! Sorry I'm on my phone so short comments lol

File: src/error.rs (hunk 4 above)
Comment (lukapeschke) [reply src/error.rs:148]: Yes but the transformation into a `PyResult` cannot fail. To me, `TryIntoPyResult` would imply

```
fn try_into_pyresult(self) -> Result<PyResult<Inner>, _>;
```

File: src/types/excelreader.rs (hunk 2 above)
Comment (lukapeschke) [reply src/types/excelreader.rs]: good catch, 0017427

File: python/fastexcel/__init__.py (hunk 1 above)
Comment (lukapeschke) [reply python/fastexcel/__init__.py:18]: Woops yes, was planning to do it and completely forgot :facepalm: 6c0a38c

File: src/error.rs (hunk 4 above)
Comment (PrettyWood) [reply src/error.rs:184]: Ahh yes yes ofc

**Review by PrettyWood — CHANGES_REQUESTED** [review]
Finally on a computer. the `make doc` command doesn't work anymore on my machine. Same for you @lukapeschke ? If it's only on my side I'll git into it

**Comment by lukapeschke** [comment]
@PrettyWood yeah docs still work on my side, see attached artifact. But I added a CI check just in case: 0c67305

[docs.zip](https://github.com/ToucanToco/fastexcel/files/14397007/docs.zip)

**Review by PrettyWood — APPROVED** [review]
Perfect! I fixed the doc on my hand sorry for the false alarm

**Pull request merged**